### PR TITLE
Latest changes for centroidal work on solo12

### PR DESCRIPTION
**Pull request by jviereck**

### CODE REVIEW

**Comment by jviereck** [comment]
Assigned to @majidkhadiv as @MaximilienNaveau is PTO.

**Comment by jviereck** [comment]
Bunch of changes.

File: python/dg_tools/leg_impedance_control/quad_leg_impedance_controller.py [hunk 1]
```diff
@@ -165,36 +165,44 @@ def __init__(self, robot, ViconClientEntity, client_name = "vicon_client" , vico
...
-            self.vicon_client.robot_wrapper(robot)
+            self.vicon_client.robot_wrapper(robot, self.robot_vicon_name)
         except:
             print("not in simulation")
 
 
-        self.robot.add_ros_and_trace(self.client_name, self.robot_vicon_name + "_position")
-        self.robot.add_ros_and_trace(self.client_name, self.robot_vicon_name + "_velocity_body")
-        self.robot.add_ros_and_trace(self.client_name, self.robot_vicon_name + "_velocity_world")
+        self.robot.add_ros_and_trace(self.vicon_client.name, self.robot_vicon_name + "_position")
```
Comment (MaximilienNaveau) [inline python/dg_tools/leg_impedance_control/quad_leg_impedance_controller.py]: Be careful with this: the export to ROS is not really real-time safe. It is better t `add_to_trace` by default and let the user define the signal he wants to export to ros on the fly.

File: python/dg_tools/leg_impedance_control/quad_leg_impedance_controller.py [hunk 2]
```diff
@@ -165,36 +165,44 @@ def __init__(self, robot, ViconClientEntity, client_name = "vicon_client" , vico
...
+        self._biased_base_position = None
+        self._biased_base_velocity = None
 
+        self.vicon_offset = constVector([0., 0., 0.,])
 
-        self.base_pos_xyz = selec_vector(self.vicon_client.signal(self.robot_vicon_name + "_position"),
-                                        0, 3, "base_pos")
+    def init_robot_properties(self):
+        self.robot_vicon_name = "quadruped"
+        self.robot_mass = [2.17784, 2.17784, 2.17784]
```
Comment (MaximilienNaveau) [inline python/dg_tools/leg_impedance_control/quad_leg_impedance_controller.py:209]: Why 3 data?

File: python/dg_tools/utils.py [hunk 3]
```diff
@@ -142,6 +305,12 @@ def component_of_vector(vector, index, entityName):
     return comp_of_vect.sout
 
 
+def sin_doub(db1, entityName=''):
```
Comment (MaximilienNaveau) [inline python/dg_tools/utils.py]: Can we have a more explicit name? `sinus_double`?

File: src/operator.cpp [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    dynamicgraph::sot::VectorQuaternion quat;
+    quat.coeffs() = input.tail<4>();
+
+    out.resize(6);
+    out.head<3>() = input.head<3>();
+    out.tail<3>() = quat.toRotationMatrix().eulerAngles(0, 1, 2);
+    return out;
+}
+
+DYNAMICGRAPH_FACTORY_ENTITY_PLUGIN(SinEntity, "Sin");
```
Comment (MaximilienNaveau) [inline src/operator.cpp]: This should be SinEntity, "SinEntity"

File: src/operator.cpp [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+/* --------------------------------------------------------------------- */
+/* --------------------------------------------------------------------- */
+/* --------------------------------------------------------------------- */
+
+SinEntity::SinEntity( const std::string & name )
```
Comment (MaximilienNaveau) [inline src/operator.cpp]: There will confusion between sin:signal_in and sin:sinus.
What about just the the `Sinus` class?
We know it comes from dg_tools so it is most likely an entity.

**Review by MaximilienNaveau — CHANGES_REQUESTED** [review]
Overall it sounds very nice, specially the operators on the signals.
Though there are some small things that needs to be addressed (See the somments).

File: python/dg_tools/leg_impedance_control/quad_leg_impedance_controller.py (hunk 2 above)
Comment (jviereck) [reply python/dg_tools/leg_impedance_control/quad_leg_impedance_controller.py:209]: Because the robot_mass is assumed to be a vector of 3.

**Comment by jviereck** [comment]
Updated to latest master.

@MaximilienNaveau , can you have a second look if you want me to change anything?

File: python/dg_tools/leg_impedance_control/quad_leg_impedance_controller.py [hunk 6]
```diff
@@ -157,43 +157,62 @@ def record_data(self, record_vicon = False):
...
+
+            self.vicon_client.add_object_to_track("{}/{}".format(self.robot_vicon_name, self.robot_vicon_name))
+
+            try:
+                ## comment out if running on real robot
+                self.vicon_client.robot_wrapper(robot, self.robot_vicon_name)
+            except:
+                print("not in simulation")
+
+            self.robot.add_ros_and_trace(self.vicon_client.name, self.robot_vicon_name + "_position")
```
Comment (MaximilienNaveau) [inline python/dg_tools/leg_impedance_control/quad_leg_impedance_controller.py]: please remove the ros here

**Review by MaximilienNaveau — APPROVED** [review]

**Comment by jviereck** [comment]
Thanks for the review Max. Addressed the `add_ros_and_trace` comment and merged.

**Pull request merged**